### PR TITLE
Don't give children negative sizes when overconstrained

### PR DESCRIPTION
**Pull request by eseidelGoogle**
We probably need a test here. Not sure how.  @Hixie 
The game was hitting this when in landscape mode.  @vlidholt

### CODE REVIEW

File: packages/flutter/lib/src/rendering/flex.dart [hunk 1]
```diff
@@ -246,7 +247,7 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
       while (child != null) {
         int flex = _getFlex(child);
         if (flex > 0) {
-          double childMainSize = spacePerFlex * flex;
+          double childMainSize = math.max(0.0, spacePerFlex * flex);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/flex.dart]: after looking at this more, I think your original intuition of having the math.max on the line that sets `spacePerFlex` is correct. It matches similar logic elsewhere.

**Comment by Hixie** [comment]
To test thing create a file that has a sized box that contains a flex that contains a sized box bigger than the parent, and then contains a flexible containing a container. Then, expect() that the container's rendor object has a height of zero.

**Comment by eseidelGoogle** [comment]
ptal

File: packages/flutter/test/rendering/flex_test.dart [hunk 2]
```diff
@@ -19,6 +19,65 @@ void main() {
     expect(flex.size.height, equals(200.0), reason: "flex height");
   });
 
+  test('Vertical Overflow', () {
+    RenderConstrainedBox flexible = new RenderConstrainedBox(
+      additionalConstraints: const BoxConstraints.expand()
+    );
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.vertical,
+      children: [
```
Comment (Hixie) [inline packages/flutter/test/rendering/flex_test.dart]: `<RenderBox>[...]`

File: packages/flutter/test/rendering/flex_test.dart [hunk 3]
```diff
@@ -19,6 +19,65 @@ void main() {
...
+    expect(flex.getMaxIntrinsicWidth(viewport), equals(100.0));
+  });
+
+  test('Horizontal Overflow', () {
+    RenderConstrainedBox flexible = new RenderConstrainedBox(
+      additionalConstraints: const BoxConstraints.expand()
+    );
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.horizontal,
+      children: [
```
Comment (Hixie) [inline packages/flutter/test/rendering/flex_test.dart]: `<RenderBox>[...]`

File: packages/flutter/test/rendering/flex_test.dart [hunk 4]
```diff
@@ -19,6 +19,65 @@ void main() {
...
+    expect(flex.getMinIntrinsicHeight(viewport), equals(100.0));
+    expect(flex.getMaxIntrinsicHeight(viewport), equals(100.0));
+    expect(flex.getMinIntrinsicWidth(viewport), equals(100.0));
+    expect(flex.getMaxIntrinsicWidth(viewport), equals(100.0));
+  });
+
+  test('Vertical Flipped Constraints', () {
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.vertical,
+      children: [
```
Comment (Hixie) [inline packages/flutter/test/rendering/flex_test.dart]: `<RenderBox>[...]`

**Comment by Hixie** [comment]
lgtm

**Pull request merged**